julia/src/native_function.jl

```diff
@@ -0,0 +1,277 @@
+    implicit::Bool
+end
+
+"Creates an `Array{T,N}``, if all elements of size are present then that array is initialized to hold `size`"
+struct ArrayBuilder{T,N}
+    size::Vector{Union{Symbol,Int}}
+end
+ArrayBuilder{T}(size) where T = ArrayBuilder{T,length(size)}(size)
+
+
+
+"""
+    NativeFunction{R}
+
+Callable Julia wrapper of some native Dex function.
+Returns type from calling it is is `R`.
+
+Usually constructed using [`@dex_func_str`](@ref),
+or via `NativeFunction(atom)` on some [`DexCall.Atom`](@ref).
+"""
+struct NativeFunction{R} <: Function
+    c_func_ptr::Ptr{Nothing}
+    implict_argument_signature::Vector{Binder}
+    explicit_argument_signature::Vector{Binder}
+    result_signature::Vector{Binder}
+end
+
+NativeFunction(atom::Atom, jit=JIT) = NativeFunction(atom.ptr, atom.ctx, jit)
+function NativeFunction(atom::Ptr{HsAtom}, ctx=PRELUDE, jit=JIT)
+    c_func_ptr = compile(ctx, atom, jit)
+    sig_ptr = get_function_signature(c_func_ptr, jit)
+    sig_ptr == C_NULL && error("Failed to retrieve the function signature")
+
+    try
+        signature = NativeFunctionSignatureJL(sig_ptr)
+        argument_signature = parse_sig(signature.arg)
+        result_signature = parse_sig(signature.res)
+        R = result_type(result_signature)
+        f = NativeFunction{R}(
+            c_func_ptr,
+            filter(x->x.implicit, argument_signature),
+            filter(x->!x.implicit, argument_signature),
+            result_signature
+        )
+
+        # We are attaching this to f.result_signature because only mutable types can have finalizers
+        finalizer(f.result_signature) do _
+            unload(c_func_ptr, jit)
+        end
+        return f
+    catch
+        unload(c_func_ptr, jit)
+        rethrow()
+    finally
+        free_function_signature(sig_ptr)
+    end
+end
+
+function (f::NativeFunction{R})(args...)::R where R
+    if length(args) != length(f.explicit_argument_signature)
+        throw(ArgumentError("wrong number of inputs, expected: $(length(f.explicit_argument_signature) )"))
+    end
+
+    named_sizes = Dict{Symbol, Int}()
+    arg_ctypes = map(f.explicit_argument_signature, args) do binder, arg
+        to_ctype!(named_sizes, binder.type, arg)
+    end
+
+    implict_values = map(f.implict_argument_signature) do binder
+        # Right now the Exported API only supports implicts that are sizes
+        # and now all of them will have been determined by the preverious `to_ctype!` step
+        @assert binder.type == Int32
+        named_sizes[binder.name]
+    end
+
+    results_ptrs_and_thunks = map(f.result_signature) do builder
+        create(builder.type, named_sizes)  
+    end
+
+    result_ptrs = first.(results_ptrs_and_thunks)
+    result_thunks = Tuple(last.(results_ptrs_and_thunks))
+
+    args = (implict_values..., arg_ctypes..., result_ptrs...)
```

Comment:
The assumption that all the implicit values come before arg_ctypes is a little sketchy and doesn't necessary have to be true. Can you follow the behavior of Python bindings and just use the full argument signature to constructor the argument tuple?

Reply:
I could.
Can you show me a case that doesn't follow this so I can add it to tests?
It was true for all the cases in the python tests.
And making that assumption simplified the code.
But I could totally do it the way python does.

Reply:
I expect that if you did this then it might fail:
```
def f (a : Float) (n : Int) ?-> (b : (Fin n)=>Float) : Float = a + sum b
```

Reply:
You are correct that that doesn't have the implict first
```
julia> DexCall.NativeFunctionSignatureJL(sig_ptr)
DexCall.NativeFunctionSignatureJL("arg0:f32,?arg1:i32,arg2:f32[arg1]", "arg3:f32", "arg0,arg1,arg2,arg3")
```

But for some reason the code works, and I can't seem to workout why
I can't be just getting lucky with pointers can I?
The following is with a show statement so I can see what is being passed to the `_dex_ccall`
```julia
julia> dex_func"def f (a : Float) (n : Int) ?-> (b : (Fin n)=>Float) : Float = a + sum b"
       f(100f0, [10f0, 2f0, 0.1f0, 0.1f0])
args = (implict_values..., arg_ctypes..., result_ptrs...) = (4, 100.0f0, Ptr{Float32} @0x00007fda2c47b590, Ptr{Nothing} @0x00007fda298e2300)
112.2f0
```

Reply:
I have fixed it anyway. Because I have no idea why that worked.
Now it only makes the assumption that return types are at the end, and that arguments go into the ccall the same order as in the the argument_signature.
Which is an assumption made in the python code here also:
https://github.com/google-research/dex-lang/blob/eb52530e3666ed1d09ac913a8edfce290548264b/python/dex/native_function.py#L107-L110

Reply:
Yes this is the right assumption to make. Our API is undocumented (of course), but those are the semantics of native function export.